### PR TITLE
distsqlrun: fix lookup join order preservation

### LINKED ISSUE #33354: sql: wrong sorting after statistics init
before statistics init:
1. explain SELECT * 
FROM "requests"  JOIN "organizations" ON (organizations.id = requests.organization_id) 
JOIN "places" AS "load_place_id_places" ON (load_place_id_places.id = requests.load_place_id) 
ORDER BY "requests"."name" desc

```
sort		
 │	order	-name
 └── join		
      │	type	inner
      │	equality	(load_place_id) = (id)
      ├── join		
      │    │	type	inner
      │    │	equality	(organization_id) = (id)
      │    ├── scan		
      │    │	table	requests@primary
      │    │	spans	ALL
      │    └── scan		
      │	table	organizations@primary
      │	spans	ALL
      └── scan		
	table	places@primary
	spans	ALL
```
2.  Statistics init:
CREATE STATISTICS places_stats ON id FROM places; 
table places has about 300k entries.

3. explain again:
```
lookup-join		
 │	type	inner
 ├── sort		
 │    │	order	-name
 │    └── join		
 │         │	type	inner
 │         │	equality	(organization_id) = (id)
 │         ├── scan		
 │         │	table	requests@primary
 │         │	spans	ALL
 │         └── scan		
 │	table	organizations@primary
 │	spans	ALL
 └── scan		
	table	places@primary
```
So after create stats we have wrong sorting.

**Environment:**
 - CockroachDB version [ 2.1.1 and 2.1.3]
 - Server OS: [Ubuntu server 18]
 - Client app [DBeaver and golang]

image
image

### PR DESCRIPTION
**Pull request by jordanlewis**
Fixes #33354.

Previously, a lookup join with a left side that doesn't have an
injective mapping into the right side would fail to preserve its input
order, in violation of its contract.

Now, the code is modified to use the order-preserving pathway that was
designed for outer joins all the time, to prevent the issue.

Release note (bug fix): lookup joins preserve their input order even if
more than one row of the input corresponds to the same row of the lookup
table.

### CODE REVIEW

**Review by RaduBerinde — COMMENTED** [review]
Thanks for fixing this so quickly!  In the issue you mention that existing logic tests would have caught this but they weren't generating the expected plan. Could we fix them (or add new versions of them)? We should be able to get this kind of plan by injecting the right statistics (note that the distinct count on the join column should be much smaller than the row count, so that we estimate that each input row produces multiple output rows).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33536#-:-LV_ipJXCCE9XIWWFhMa:b7kjgfz)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by jordanlewis — COMMENTED** [review]
I did modify the logic test so that it fails without the patch - I 10x'd the row count. It should be in the diff to the `lookup_join` logic test. I was thinking I could also add the exact minimal regression test that I posted in the issue, but it's kind of a duplicate - what do you think?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33536#-:-LV_n2Ln2CyjbQbAKy0L:b5fvkry)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by RaduBerinde — COMMENTED** [review]
Oh, I missed that. Can you also update the corresponding test in `execbuilder/testdata/lookup_join`? That should contain the corresponding plans. (btw looks like the injected stats in this file have been changed and haven't been kept in sync with the other one, they should be the same)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33536#-:-LV_oa-16aNVhNU8mM_9:bz35oaz)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by jordanlewis — COMMENTED** [review]
Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33536#-:-LVa0gZi6HLv_NdZGgh7:b-896fix)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by RaduBerinde — COMMENTED** [review]
I'm not too familiar with the joinreader code so you might want to wait for another review on that. Otherwise :lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33536#-:-LVczieMEJNMK9hy4Qpq:bbrkq8z)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/opt/exec/execbuilder/testdata/lookup_join, line 225 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33536#-LVcz_lW1AwNm_PIbPxS:-LVcz_lW1AwNm_PIbPxT:b-nrtj59) ([raw file](https://github.com/cockroachdb/cockroach/blob/53024918d0ffdc90c5cc76b1ce81d88727cfa9ef/pkg/sql/opt/exec/execbuilder/testdata/lookup_join#L225)):*
> ```
> SELECT url FROM [EXPLAIN (DISTSQL) SELECT DISTINCT authors.name FROM books AS b1, books2 AS b2, authors WHERE b1.title = b2.title AND authors.book = b1.title AND b1.shelf <> b2.shelf]
> ----
> https://cockroachdb.github.io/distsqlplan/decode.html#eJzEk02L2zAQhu_9FeqcElCx5a8FQ0CFPXR78Jalt5KDYk0TUccykgwtIf-92IYmdmPl47I3aTTPvDPvoAPUWmIh9mgh_wEMKKSwptAYXaK12nThIelF_oY8pKDqpnVdeE2h1AYhP4BTrkLIodCfdBNkQEGiE6rq044UdOtOkHVii5A_HelZYeYv_F1sKnxDIdEE4ag8NEbthfnDN1r_shFQeG1dTjijPIY5bfaoNrusLVq367w6E49mxaNZ8ZNmW2sj0aCcWnk95cIEX4TdfdWqRhNE4wFO7VIeU57MNh0_6ljs2dZty0re069k3H6FP92Cs-XKqO2uP3UzFGTBE7IiPF2Sz8UzWfCMfFwRHi3_TTg_XnqPs8_KOlWXLkjHfXE2Wz8b1b_ygd_QNrq2eNMPDjvvUG5x2IXVrSnxm9FlLzNcX3uuD0i0bnh9Gi4v9fDUNXgOMy8cj2A2haM74GgKx1449Ssnd8D_KadeOPMblnnhcAKvjx_-BgAA__-eevL0
> ```

Not this change but this doesn't seem to be a lookup join. Can you add a `TODO(radu)` to investigate?

**Review by solongordon — APPROVED** [review]
:lgtm: Thanks for the fix. Frustratingly I think this is how I originally wrote the ordering fix for left joins, but then I special cased it because I was trying to avoid unnecessary work.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33536#-:-LVsHedr1Cmv7C_KAkhi:b-4nimny)* status: :shipit: complete! 2 of 0 LGTMs obtained

**Pull request merged**
